### DISCUSSION TITLE
Possibility to set/override checkout field values programmatically

### DISCUSSION THREAD

**Discussion by Dandev85** [woocommerce/woocommerce#50119] [WooCommerce, Blocks, and Site Editing]
Hello,

I am trying to adapt my plugin in order to make it compatible with the new Checkout Block.

With our plugin a customer can place orders on behalf of other customers and we use the woocommerce_checkout_get_value hook to override checkout field values, e.g.:

```
function set_checkout_field_values_programmatically( $value, $input ) {
	
	$user_id = 123;

	$user = new WC_Customer( $user_id );

	// Some conditions...
	if ( ... ) {
		
		$checkout_fields = array(
		
			// Billing details.
			'billing_first_name'    => $user->get_billing_first_name(),																						
			'billing_last_name'     => $user->get_billing_last_name(),
			'billing_company'       => $user->get_billing_company(),
			'billing_country'       => $user->get_billing_country(),
			'billing_address_1'     => $user->get_billing_address_1(),
			'billing_address_2'     => $user->get_billing_address_2(),
			'billing_city'          => $user->get_billing_city(),
			'billing_state'         => $user->get_billing_state(),
			'billing_postcode'      => $user->get_billing_postcode(),
			'billing_phone'         => $user->get_billing_phone(),
			'billing_email'         => $user->get_billing_email(),
			
			// Shipping details.
			'shipping_first_name'   => $user->get_shipping_first_name(),
			'shipping_last_name'    => $user->get_shipping_last_name(),
			'shipping_company'      => $user->get_shipping_company(),
			'shipping_country'      => $user->get_shipping_country(),
			'shipping_address_1'    => $user->get_shipping_address_1(),
			'shipping_address_2'    => $user->get_shipping_address_2(),
			'shipping_city'         => $user->get_shipping_city(),
			'shipping_state'        => $user->get_shipping_state(),
			'shipping_postcode'     => $user->get_shipping_postcode(),
		);
	}
	
	if ( ! empty( $checkout_fields ) ) {
		
		// Populate checkout fileds with custom billing and shipping details.
		foreach( $checkout_fields as $key_field => $field_value ){
			if( $input == $key_field ){
				$value = $field_value;
			}
		}
	}
		
	return $value;
}
add_filter( 'woocommerce_checkout_get_value', 'set_checkout_field_values_programmatically', 10, 2 );
```
There may also be other scenarios where you need to set/override the default checkout field values (https://stackoverflow.com/questions/45602936/woocommerce-set-checkout-field-values).

So I was wondering if you are planning to provide the possibility to set/override customer checkout field values within the new Checkout Block too.

Thank you in advance,
Dan

**Comment by senadir** [comment]
Hey Dan! `woocommerce_checkout_get_value` seems to be ephemeral and would only last for the callback duration, would setting the value directly in the customer object not work for you?

**Comment by Dandev85** [comment]
Hi and thank you for your reply!

I'll share a few more details to provide more context regarding our goal.

Let's say we have two customers, **Customer X** and **Customer Y**.

_Customer X_ has its own individual billing/shipping details.

_Customer Y_ also has its own individual billing/shipping details (different from Customer X).

With our plugin, _Customer X_ can:

- place orders for himself
- place orders on behalf of _Customer Y_

Now, (only) when _Customer X_ places an order on behalf of _Customer Y_, this is what happens:

1. we do show billing/shipping details of "Customer Y" in the checkout form (checkout fields are readonly and cannot be modified in this case)
2. we save billing/shipping details of "Customer Y" in that particular order (in the order metadata)
3. we prevent WooCommerce from updating user's billing/shipping address information from the checkout form: `add_filter( 'woocommerce_checkout_update_customer_data', '__return_false' );`

Point 1 is quite crucial for us, because it warns _Customer X_ that he is placing an order on behalf of _Customer Y_.

With the shortcode-based checkout, this is achieved using the function provided in my previous message, but unfortunately with the Checkout Block we are not able to achieve the same result.

Regarding point 1, could there be a different approach to achieve the same result using the Checkout Block or are we stuck with the shortcode-based checkout method?

Thank you in advance,
Dan

**Comment by senadir** [comment]
Hey @Dandev85 

I'm still thinking about how to best support you here. For now, you might consider doing things like switching up the value of `wc()->customer` depending on who you want to show.

**Comment by Dandev85** [comment]
Hello @senadir 

Thank you for your suggestion, `WC()->customer->set_...` works perfectly for our needs. Your help is much appreciated.

I see that `add_filter( 'woocommerce_checkout_update_customer_data', '__return_false' );` no longer works when using the checkout block.

As a workaround, for now we are storing the original user details in a WC session (and restoring them by hooking into `woocommerce_thankyou`).

Before proceeding this way, I wanted to know if you are planning to provide a more official/consistent way to prevent WooCommerce from updating user's billing/shipping address information after the checkout process (when using the checkout block).

Thank you in advance!

**Comment by andrewklimek** [comment]
I don't use block checkout, but it would definitely be good to get 'woocommerce_checkout_update_customer_data' working in there in case I ever do
